test/common/stats/stat_test_utility.h

```diff
@@ -72,6 +74,45 @@ class MemoryTest {
   const size_t memory_at_construction_;
 };
 
+// Helps tests construct StatName with symbolic and dynamic components.
+class MixedStatNames {
```

Comment:
My main concern with this type of construct is whether test writers are going to understand that they need to do this. Is there any way that we could make this fail in a more obvious way if a test writer is not doing the right thing? Possibly by making the test stat stores use a slightly different interface? I'm not sure what the best option is.

Reply:
I will address the other comments  a little later, but this is the most strategic question.

After this PR, a dev writing a unit-test that looks for a counter in a map, and has used a Dynamic stat name, will encounter a test failure, and will have to discover how to use this solution. More hints could definitely be dropped in the doc doc for StatNameDynamicPool and stats.md. That would clearly be in-scope for this PR.

Another alternative, which maybe I should've considered more carefully, is to add a string-based map to a test override for IsolatedStatsStore. A warning could be placed in IsolatedStatsStore doc to use that instead. The advantage of this approach is that no one has to go crazy with the backquotes :)

The disadvantage is that if someone writes an integration test for a subsystem that uses dynamic stats, using the production stats stores, they will still need to do something manual to make stat lookups by name work.

Reply:
WDYT of having a separate effort in parallel to migrate all tests away from the pattern of using counter("foo").value() and toward one of instantiating something a new class StatTestContext, which could build its own string->stat map on demand. Once the string-map is populated, no distinction is needed between dynamic and static counters.

There are O(1500) references to pull counters and gauges from test files, so it would be an annoying PR but it wouldn't be too bad.

Then we could remove the deprecated counter() and gauge() methods from the Stats interface.

WDYT also of doing that as a follow-up to this PR, which is needed to to fix a bug?

Reply:
+1 to your last comment, I think that is the right long term solution and we should just bite the bullet and do the cleanup. Otherwise I fear that this is going to confuse people in very hard to understand ways. Is there anything we can do further in this PR doc wise that we can do to help people until that is done?

/wait-any

Reply:
Yes, will update doc for StatNameDynamicStorage and StatNameDynamicPool to warn about lookups for testing, and also stats.md. SG?

I'll also revert the change in symbol_table_creator.cc so that fakes remain the default for testing.

In parallel, in a PR branched from this one, I'll try to start the slog of changing all tests to reference a helper class for stat lookup by name.

Reply:
OK I have an alternative to this one in #9836 -- no backquotes needed. A test-specific structure is created that wraps a Store&, tries to keep string-maps lazily, and thus provides test-specific counter lookups that are independent of how the StatName was constructed.

I'm not happy with the new class name in that PR yet, and there's about 60+ more files that need to be converted over. But they don't need to be done all at once. I could also add a format-checker that whitelists the current usage until we can remove the calls entirely. One other complication to getting all that done is how to handle mocks.